# In-Class Coding Lab: Data Analysis with Pandas

In this lab, we will perform a data analysis on the **RMS Titanic** passenger list. The RMS Titanic is one of the most famous ocean liners in history. On April 15, 1912 it sank after colliding with an iceberg in the North Atlantic Ocean. To learn more, read here: https://en.wikipedia.org/wiki/RMS_Titanic 

Our goal today is to perform a data analysis on a subset of the passenger list. We're looking for insights as to which types of passengers did and didn't survive. Women? Children? 1st Class Passengers? 3rd class? Etc. 

I'm sure you've heard the expression often said during emergencies: "Women and Children first" Let's explore this data set and find out if that's true!

Before we begin you should read up on what each of the columns mean in the data dictionary. You can find this information on this page: https://www.kaggle.com/c/titanic/data 


## Loading the data set

First we load the dataset into a Pandas `DataFrame` variable. The `sample(10)` method takes a random sample of 10 passengers from the data set.

In [1]:
import pandas as pd
import numpy as np

# this turns off warning messages
import warnings
warnings.filterwarnings('ignore')

passengers = pd.read_csv('CCL-titanic.csv')
passengers.sample(10)

PassengerId  Survived  Pclass                          Name     Sex  \
277          278         0       2   Parkes, Mr. Francis "Frank"    male   
516          517         1       2  Lemore, Mrs. (Amelia Milley)  female   
210          211         0       3                Ali, Mr. Ahmed    male   
606          607         0       3             Karaic, Mr. Milan    male   
499          500         0       3            Svensson, Mr. Olof    male   
303          304         1       2           Keane, Miss. Nora A  female   
124          125         0       1   White, Mr. Percival Wayland    male   
551          552         0       2   Sharp, Mr. Percival James R    male   
538          539         0       3      Risien, Mr. Samuel Beard    male   
331          332         0       1           Partner, Mr. Austen    male   

      Age  SibSp  Parch              Ticket     Fare Cabin Embarked  
277   NaN      0      0              239853   0.0000   NaN        S  
516  34.0      0      0          C.A. 34260  10.5000   F33        S  
210  24.0      0      0  SOTON/O.Q. 3101311   7.0500   NaN        S  
606  30.0      0      0              349246   7.8958   NaN        S  
499  24.0      0      0              350035   7.7958   NaN        S  
303   NaN      0      0              226593  12.3500  E101        Q  
124  54.0      0      1               35281  77.2875   D26        S  
551  27.0      0      0              244358  26.0000   NaN        S  
538   NaN      0      0              364498  14.5000   NaN        S  
331  45.5      0      0              113043  28.5000  C124        S

## How many survived?

One of the first things we should do is figure out how many of the passengers in this data set survived. Let's start with isolating just the `'Survivied'` column into a series:

In [2]:
passengers['Survived'].sample(10)

448    1
862    1
857    1
555    0
306    1
0      0
890    0
742    1
374    0
369    1
Name: Survived, dtype: int64

There's too many to display so we just display a random sample of 10 passengers. 

- 1 means the passenger survivied
- 0 means the passenger died

What we really want is to count the number of survivors and deaths. We do this by querying the `value_counts()` of the `['Survived']` column, which returns a `Series` of counts, like this:

In [3]:
passengers['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

Only 342 passengers survived, and 549 perished. Let's observe this same data as percentages of the whole. We do this by adding the `normalize=True` named argument to the `value_counts()` method.

In [4]:
passengers['Survived'].value_counts(normalize=True)

0    0.616162
1    0.383838
Name: Survived, dtype: float64

**Just 38% of passengers in this dataset survived.**

### Now you Try it!

**FIRST** Write a Pandas expression to display counts of males and female passengers using the `Sex` variable:

In [5]:
# todo write code here
passengers['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

**NEXT** Write a Pandas expression to display male /female passenger counts as a percentage of the whole number of passengers in the data set.

In [6]:
# todo write code here
passengers['Sex'].value_counts(normalize=True)

male      0.647587
female    0.352413
Name: Sex, dtype: float64

If you got things working, you now know that **35% of passengers were female**.

## Who survivies? Men or Women?

We now know that 35% of the passengers were female, and 65% we male. 

**The next think to think about is how do survivial rates affect these numbers? **

If the ratio is about the same for surviviors only, then we can conclude that your **Sex** did not play a role in your survival on the RMS Titanic. 

Let's find out.

In [7]:
survivors = passengers[passengers['Survived'] ==1]
survivors['PassengerId'].count()

342

Still **342** like we discovered originally. Now let's check the **Sex** split among survivors only:

In [8]:
survivors['Sex'].value_counts()

female    233
male      109
Name: Sex, dtype: int64

WOW! That is a huge difference! But you probably can't see it easily. Let's represent it in a `DataFrame`, so that it's easier to visualize:

In [9]:
sex_all_series = passengers['Sex'].value_counts()
sex_survivor_series = survivors['Sex'].value_counts()

sex_comparision_df = pd.DataFrame({ 'AllPassengers' : sex_all_series, 'Survivors' : sex_survivor_series })
sex_comparision_df['SexSurvivialRate'] = sex_comparision_df['Survivors'] / sex_comparision_df['AllPassengers']
sex_comparision_df

AllPassengers  Survivors  SexSurvivialRate
female            314        233          0.742038
male              577        109          0.188908

 **So, females had a 74% survival rate. Much better than the overall rate of 38%**
 
We should probably briefly explain the code above. 

- The first two lines get a series count of all passengers by Sex (male / female) and count of survivors by sex
- The third line creates DataFrame. Recall a pandas dataframe is just a dict of series. We have two keys 'AllPassengers' and 'Survivors'
- The  fourth line creates a new column in the dataframe which is just the survivors / all passengers to get the rate of survival for that Sex.

## Feature Engineering: Adults and Children

Sometimes the variable we want to analyze is not readily available, but can be created from existing data. This is commonly referred to as **feature engineering**. The name comes from machine learning where we use data called *features* to predict an outcome. 

Let's create a new feature called `'AgeCat'` as follows:

- When **Age** <=18 then 'Child'
- When **Age** >18 then 'Adult'

This is easy to do in pandas. First we create the column and set all values to `np.nan` which means 'Not a number'. This is Pandas way of saying no value. Then we set the values based on the rules we set for the feature.

In [10]:
passengers['AgeCat'] = np.nan # Not a number
passengers['AgeCat'][ passengers['Age'] <=18 ] = 'Child'
passengers['AgeCat'][ passengers['Age'] > 18 ] = 'Adult'
passengers.sample(5)

PassengerId  Survived  Pclass                         Name     Sex   Age  \
83            84         0       1      Carrau, Mr. Francisco M    male  28.0   
631          632         0       3  Lundahl, Mr. Johan Svensson    male  51.0   
215          216         1       1      Newell, Miss. Madeleine  female  31.0   
865          866         1       2     Bystrom, Mrs. (Karolina)  female  42.0   
335          336         0       3           Denkoff, Mr. Mitto    male   NaN   

     SibSp  Parch  Ticket      Fare Cabin Embarked AgeCat  
83       0      0  113059   47.1000   NaN        S  Adult  
631      0      0  347743    7.0542   NaN        S  Adult  
215      1      0   35273  113.2750   D36        C  Adult  
865      0      0  236852   13.0000   NaN        S  Adult  
335      0      0  349225    7.8958   NaN        S    NaN

Let's get the count and distrubutions of Adults and Children on the passenger list.

In [11]:
passengers['AgeCat'].value_counts()

Adult    575
Child    139
Name: AgeCat, dtype: int64

And here's the percentage as a whole:

In [12]:
passengers['AgeCat'].value_counts(normalize=True)

Adult    0.805322
Child    0.194678
Name: AgeCat, dtype: float64

So close to **80%** of the passengers were adults. Once again let's look at the ratio of `AgeCat` for survivors only. If your age has no bearing of survivial, then the rates should be the same. 

Here's the counts of Adult / Children among the survivors only:

In [13]:
survivors = passengers[passengers['Survived'] ==1]
survivors['AgeCat'].value_counts()

Adult    220
Child     70
Name: AgeCat, dtype: int64

### Now You Try it!

Calculate the `AgeCat` survival rate, similar to how we did for the `SexSurvivalRate`. 

In [14]:
agecat_all_series = passengers['AgeCat'].value_counts()
agecat_survivor_series = survivors['AgeCat'].value_counts()

agecat_survival_rate_df = pd.DataFrame({ 'AllPassengers' : agecat_all_series, 'Survivors' : agecat_survivor_series })
agecat_survival_rate_df['survivalrate'] = agecat_survival_rate_df['Survivors'] / agecat_survival_rate_df['AllPassengers']
agecat_survival_rate_df

# todo make a data frame, add AgeCatSurvivialRate column, display dataframe 



AllPassengers  Survivors  survivalrate
Adult            575        220      0.382609
Child            139         70      0.503597

In [15]:
sex_all_series = passengers['Sex'].value_counts()
sex_survivor_series = survivors['Sex'].value_counts()

sex_comparision_df = pd.DataFrame({ 'AllPassengers' : sex_all_series, 'Survivors' : sex_survivor_series })
sex_comparision_df['SexSurvivialRate'] = sex_comparision_df['Survivors'] / sex_comparision_df['AllPassengers']
sex_comparision_df

AllPassengers  Survivors  SexSurvivialRate
female            314        233          0.742038
male              577        109          0.188908

**So, children had a 50% survival rate, better than the overall rate of 38%**

## So, women and children first?

It looks like the RMS really did have the motto: "Women and Children First."

Here's our insights. We know:

- If you were a passenger, you had a 38% chance of survival.
- If you were a female passenger, you had a 74% chance of survival.
- If you were a child passenger, you had a 50% chance of survival. 


### Now you try it for Passenger Class

Repeat this process for `Pclass` The passenger class variable. Display the survival rates for each passenger class. What does the information tell you about passenger class and survival rates?

I'll give you a hint... "Money Talks"


In [16]:
# todo: repeat the analysis in the previous cell for Pclass 
pclass_all_series = passengers['Pclass'].value_counts()
pclass_survivor_series =survivors['Pclass'].value_counts()

pclass_compare_df = pd.DataFrame({'AllPassengers' :pclass_all_series, 'Survivors' :pclass_survivor_series })
pclass_compare_df['Pclass'] = pclass_compare_df['Survivors'] / pclass_compare_df['AllPassengers']
pclass_compare_df

AllPassengers  Survivors    Pclass
1            216        136  0.629630
2            184         87  0.472826
3            491        119  0.242363

In [16]:
sex_all_series = passengers['Sex'].value_counts()
sex_survivor_series = survivors['Sex'].value_counts()

sex_comparision_df = pd.DataFrame({ 'AllPassengers' : sex_all_series, 'Survivors' : sex_survivor_series })
sex_comparision_df['SexSurvivialRate'] = sex_comparision_df['Survivors'] / sex_comparision_df['AllPassengers']
sex_comparision_df

AllPassengers  Survivors  SexSurvivialRate
female            314        233          0.742038
male              577        109          0.188908

In [15]:
passengers = pd.read_csv('CCL-titanic.csv')
passengers.sample(10)


PassengerId  Survived  Pclass  \
379          380         0       3   
704          705         0       3   
608          609         1       2   
19            20         1       3   
696          697         0       3   
745          746         0       1   
371          372         0       3   
423          424         0       3   
475          476         0       1   
452          453         0       1   

                                                  Name     Sex   Age  SibSp  \
379                        Gustafsson, Mr. Karl Gideon    male  19.0      0   
704                            Hansen, Mr. Henrik Juul    male  26.0      1   
608  Laroche, Mrs. Joseph (Juliette Marie Louise La...  female  22.0      1   
19                             Masselmani, Mrs. Fatima  female   NaN      0   
696                                   Kelly, Mr. James    male  44.0      0   
745                       Crosby, Capt. Edward Gifford    male  70.0      1   
371                          Wiklund, Mr. Jakob Alfred    male  18.0      1   
423  Danbom, Mrs. Ernst Gilbert (Anna Sigrid Maria ...  female  28.0      1   
475                        Clifford, Mr. George Quincy    male   NaN      0   
452                    Foreman, Mr. Benjamin Laventall    male  30.0      0   

     Parch         Ticket     Fare Cabin Embarked  
379      0         347069   7.7750   NaN        S  
704      0         350025   7.8542   NaN        S  
608      2  SC/Paris 2123  41.5792   NaN        C  
19       0           2649   7.2250   NaN        C  
696      0         363592   8.0500   NaN        S  
745      1      WE/P 5735  71.0000   B22        S  
371      0        3101267   6.4958   NaN        S  
423      1         347080  14.4000   NaN        S  
475      0         110465  52.0000   A14        S  
452      0         113051  27.7500  C111        C